llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2309 @@
+    return Symbol;
+
+  if (!options().getPrintSymbols()) {
+    // Just create the symbol type.
+    getOrCreateType(Var->getType());
+    if (const DIGlobalVariable *GV = dyn_cast<DIGlobalVariable>(Var)) {
+      if (MDTuple *TP = GV->getTemplateParams())
+        addTemplateParams(Symbol, DINodeArray(TP));
+    }
+    return nullptr;
+  }
+
+  if (!Symbol)
+    Symbol = static_cast<LVSymbol *>(constructElement(Var));
+  if (Symbol && !Symbol->getIsFinalized()) {
+    Symbol->setIsFinalized();
+    LVScope *Parent = getParentScope(Var);
+    Parent->addElement(Symbol);
+
+    Symbol->setName(Var->getName());
+
+    // Create symbol type.
+    if (LVElement *SymbolType = getOrCreateType(Var->getType()))
+      Symbol->setType(SymbolType);
+
+    if (const DILocalVariable *LV = dyn_cast<DILocalVariable>(Var)) {
+      // Add line number info.
+      addSourceLine(Symbol, LV);
+      if (LV->isParameter()) {
+        Symbol->setIsParameter();
+        if (LV->isArtificial())
+          Symbol->setIsArtificial();
+      }
+    } else {
+      const DIGlobalVariable *GV = dyn_cast<DIGlobalVariable>(Var);
+      if (useAllLinkageNames())
+        Symbol->setLinkageName(GV->getLinkageName());
+
+      // Get 'declaration' node in order to generate the DW_AT_specification.
+      if (const DIDerivedType *GVDecl = GV->getStaticDataMemberDeclaration()) {
+        LVSymbol *Reference = static_cast<LVSymbol *>(getOrCreateType(GVDecl));
+        if (Reference) {
+          Symbol->setReference(Reference);
+          Symbol->setHasReferenceSpecification();
+        }
+      } else {
+        if (!GV->isLocalToUnit())
+          Symbol->setIsExternal();
+        // Add line number info.
+        addSourceLine(Symbol, GV);
+      }
+
+      if (MDTuple *TP = GV->getTemplateParams())
+        addTemplateParams(Symbol, DINodeArray(TP));
+    }
+  }
+
+  // Create the 'inlined' symbol.
+  if (DL)
+    getOrCreateInlinedVariable(Symbol, DL);
+
+  return Symbol;
+}
+
+#ifdef LLVM_DEBUG
+void LVIRReader::printAllInstructions(BasicBlock *BB) {
+  LLVM_DEBUG({
+    dbgs() << "\nBegin all instructions:\n";
+    for (Instruction &I : *BB) {
+      dbgs() << "Instruction: '" << I << "'\n";
+      for (DbgVariableRecord &DVR : filterDbgVars(I.getDbgRecordRange()))
+        DVR.getVariable()->dump();
+    }
+    dbgs() << "End all instructions:\n\n";
+  });
+}
+#endif
+
+void LVIRReader::processBasicBlocks(Function &F, const DISubprogram *SP) {
+  assert(SP && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[processBasicBlocks] DISubprogram\n";
+    SP->dump();
+  });
+
+  // Check if we need to add a dwarf::DW_TAG_unspecified_parameters.
+  bool AddUnspecifiedParameters = false;
+  if (const DISubroutineType *SPTy = SP->getType()) {
+    DITypeRefArray Args = SPTy->getTypeArray();
+    unsigned N = Args.size();
+    if (N > 1) {
+      const DIType *Ty = Args[N - 1];
+      if (!Ty)
+        AddUnspecifiedParameters = true;
```

Comment:
Ditto (about `DW_TAG_unspecified_parameters`) -- I am not sure about this one though.